---
title: "bugphyzz"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{bugphyzz}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message=FALSE}
library(bugphyzz)
library(purrr)
library(tibble)
```

## Introduction

This document shows the main functions and operations that can be used
to work with the bugphyzz datasets, such as seeing all physiologies, 
look at specific physiologies, and make signature sets. Bugphyzz is a collection of physiological features about bacteria. We are curating the databases through 3 main methods: webscraping, bulk download from online databases, and manual curation from literature. Mainly this database will be used to analyze microbiome data.

## Definitions
* **NCBI ID**: National Center for Bioinfomatic numeric ID to link to specific taxa.
* **Genome ID**: From NCBI, another type of numeric ID, that links to a specific sequenced genome. Not all taxa have a NCBI ID, and might only have a Genome ID
* **Accession ID**: The accession number is a unique identifier assigned to a record in sequence databases such as GenBank. Several NCBI databases use the format [alphabetical prefix][series of digits]. A change in the record in some databases (e.g. GenBank) is tracked by an integer extension of the accession number, an Accession.
* **Taxon name**: a human readable organismal name that can delinate which level taxa the annotation is describing.
* **Attribute**: a value for a physiology, that is mapped to an ontology whenever possible. These are defined in the `attribute` column of [inst/extdata/attributes.tsv](https://github.com/waldronlab/bugphyzz/blob/main/inst/extdata/attributes.tsv
 **Physiology**: a microbial characteristic that can be measured or observed, that should be mapped to an ontology wherever possible. These are defined in [inst/extdata/links.tsv](https://github.com/waldronlab/bugphyzz/blob/main/inst/extdata/links.tsv).
* **Attribute value**: a value describing a physiology. Allowable values are defined by the `validity` column of [inst/extdata/attributes.tsv](https://github.com/waldronlab/bugphyzz/blob/main/inst/extdata/attributes.tsv). For categorical attributes (such as "obligately aerobic"), values can be "TRUE" and "FALSE", for example. For numeric attributes, these can be numeric values.
* **Attribute source**: a citeable domain which the annotation was obtained. A source can be another database or primary literature.
* **Frequency**: Describes how often the attritbute (phenotype) occurs: Rarely, Sometimes, Usually, Always, or Unknown. If the source makes the attribute unclear how often it occurs, then an unknown is given.
* **Evidence codes**: As the Gene Ontology defines different forms of evidence a source was obtained the annotation. 

EXP: Experimental data, the attribute was obtained through labortory confirmations, and ideally more than once.

IGC: Computational predicted through means of metagenomic analysis, or more broadly, genome analysis.

ASR: Anstrestal State Reconstruction, was predicted using ASR computional techniques to predicted up and down the taxonmic tree of taxa that has a value.

* **Note**: Any additional information that does not fit in any fixed column. The information here is variable.
* **MIDAS ID**: Located in the Habitat attribute. Any annotation from the MIDAS database is assigned an ID number. We provided that ID here.
* **Parent name**: The parent of descendant given in Taxon name column. Values can indudes names or NA.
* **Species name**: The species name, or descendant of parent, given in the Taxon name column. Values can indudes names or NA.
* **Unit**: defines unit of attribute, like Âµm.
* **Rank**: The taxonomic rank of the annotation. Values can indudes names or NA.
* **Parent Rank**: The taxonmic rank of the parent of the annotation. Values can indudes names or NA.

## Installation

You can install the development version of bugphyzz with:

``` r
# install.packages("devtools")
devtools::install_github("waldronlab/bugphyzz")
```

## Some examples of how to create signatures

``` r
suppressMessages({
    library(bugphyzz)
    library(purrr)
    aer <- physiologies('aerophilicity')[[1]]
    ph <- physiologies('optimal ph')[[1]]
})

## Example 1
## Create a signature of genera with experimental evidence at a 
## Frequency of always using the NCBI taxids. Minimumn of 10 microbes per
## signature
aer_sig <- makeSignatures(
    aer, tax.id.type = 'NCBI_ID', tax.level = 'genus', 
    Frequency = 'always', Evidence = 'exp', min.size = 10
)
#> Warning: Dropping 3648 rows with FALSE values.
#> Creating signature for categorical attributes.

map_int(aer_sig, length)
#>                 aerobic               anaerobic facultatively anaerobic 
#>                     574                     294                     174 
#>         microaerophilic      obligately aerobic    obligately anaerobic 
#>                      17                      52                     104
map(aer_sig, ~ head(.x, 2))
#> $aerobic
#> [1]  291967 1078830
#> 
#> $anaerobic
#> [1] 1427378   33951
#> 
#> $`facultatively anaerobic`
#> [1]  2147 12914
#> 
#> $microaerophilic
#> [1] 46352 12960
#> 
#> $`obligately aerobic`
#> [1] 442869    434
#> 
#> $`obligately anaerobic`
#> [1] 258514  31980


## Example 2
## Create a signature of acidiphilic species using the Taxon names
acid_sig <- makeSignatures(
    df = ph, tax.id.type = 'Taxon_name', tax.level = 'species',
    max = 4
)
#> Creating signature for continuous attribute.
head(acid_sig[[1]])
#> [1] "Leptospirillum ferriphilum"     "Sulfobacillus acidophilus"     
#> [3] "Acidithiobacillus ferrivorans"  "Acidithiobacillus ferrooxidans"
#> [5] "Acidimicrobium ferrooxidans"    "Thermoplasma volcanium"

## Create a signature of Alkaliphilic species using the Taxon names
basic_sig <- makeSignatures(
    df = ph, tax.id.type = 'Taxon_name', tax.level = 'species', 
    min = 8
)
#> Creating signature for continuous attribute.
head(basic_sig[[1]])
#> [1] "Arhodomonas aquaeolei"        "Glaciecola nitratireducens"  
#> [3] "Pseudomonas coronafaciens"    "Sodalis glossinidius"        
#> [5] "Photorhabdus luminescens"     "Hyphomicrobium nitrativorans"
```

<sup>Created on 2022-09-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

## bugphyzzAnalyses

Some analyses using data provided by this package can be found at https://waldronlab.io/bugphyzzAnalyses/

## How to add a new attribute in GitHub to the BugPhyzz database
 1. Add attributes of physiology to Attributes using ontology terms
 2. Add reference(s) to Confidence in Curation
 3. Add Google Doc "publish to web" link and add in Links
 4. Add Google Doc link to Source Link


## Installation

Bugphyzz can be installed with:

```{r, eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager") 

BiocManager::install("waldronlab/bugphyzz")
```

## Importing datasets

The main function for importing datasets is the `physiologies` function, which
has a single argument, `keyword`, and returns a list of data.frames.
The `keyword` argument accepts a character vector specifying the name(s) 
of the dataset(s) that will be imported.

A datset with categorical attribute values:

```{r, message=FALSE}
aer <- as_tibble(physiologies("aerophilicity")[[1]])
aer
```

A dataset with numeric attribute values:

```{r}
ph <- as_tibble(physiologies('optimal ph')[[1]])
ph
```

The `physiologiesList` function can be used to display which datasets are
available through the `physiologies` function before importing them:

```{r}
physiologiesList()
```

You can use the `keyword = "all"` to import all of the datasets:

```{r, message=FALSE, eval=FALSE}
## Import all of the datasets available through the physiologies function
phys <- map(physiologies(keyword = "all"), as_tibble)
```

A separate function is provided to import the fatty acid compositions dataset,
`fattyAcidComposition`. This function takes no arguments and returns a single
data.frame object:

```{r}
fac <- as_tibble(fattyAcidComposition())
fac
```

## Creating microbial signatures

Bugphyzz provides the `makeSignatures` function to create a list of microbial
signatures from a bugphyzz dataset.

### Create signatures of an attribute with categorical/logical values:

Create a list of signatures with the NCBI taxids of microbes at the
taxonomic rank of genus. Only include microbes annotated based on
experimental data with a high frequency value. The minimum number of
microbes per signature is 10.

```{r}
aer_sig <- makeSignatures(
    aer, tax.id.type = 'NCBI_ID', tax.level = 'genus', 
    Frequency = 'always', Evidence = 'exp', min.size = 10
)
map(aer_sig, ~ head(.x, 2))
```

Number of microbes per signature:

```{r}
map_int(aer_sig, length)
```

### Create signatures of an atttribute with numerical values

Create a signature list with species of acidiphilic microbes (ph <= 4):

```{r}
acid_sig <- makeSignatures(
    df = ph, tax.id.type = 'Taxon_name', tax.level = 'species',
    max = 4
)
map(acid_sig, ~ head(.x, 2))
```

Number of microbes:

```{r}
map_int(acid_sig, length)
```

Create a signature list with species of alkaliphilic microbes (ph >= 8):

```{r}
basic_sig <- makeSignatures(
    df = ph, tax.id.type = 'Taxon_name', tax.level = 'species', 
    min = 8
)
map(basic_sig, ~ head(.x, 2))
```

Number of microbes:

```{r}
map_int(basic_sig, length)
```

## Session information:

```{r}
sessionInfo()
```
