---
title: "Example of Getting Abundance Values for Categorical Attributes (Aerophilicity)"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Example of Getting Abundance Values for Categorical Attributes (aerophilicity)}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

```{r , include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  message = FALSE,
  warning = FALSE
)
```


# Intro

In this vignette, we'll use a couple of data sets (one from cMD3 and the
other from HMP16SData) to exemplify how to combine relative abundance values
with a bugphyzz attribute (aerophilicity).

We'll use the `taxize` package to assign NCBI taxids to the features (genera).

Packages:

```{r setup, message=FALSE, warning=FALSE}
library(bugphyzz)
suppressPackageStartupMessages(library(HMP16SData))
suppressPackageStartupMessages(library(curatedMetagenomicData))
library(dplyr)
library(magrittr)
library(tidyr)
library(taxize)
library(ggplot2)
```

Load the aerophilicity data set from bugphyzz:

```{r, message=FALSE}
aerophilicity <- physiologies(keyword = "aerophilicity")[[1]] %>% 
  tibble::as_tibble()
aerophilicity
```

# AsnicarF 2017 from curatedMetagenomicData (cMD)

Let's start by loading the relative abundance values from the AsnicarF_2017 data
set from cMD3. These data are stored in a SummarizedExperiment-class object.

```{r, message=FALSE, warning=FALSE}
asnicarf_se <- curatedMetagenomicData("AsnicarF_2017.relative_abundance",
                       dryrun = FALSE)
asnicarf_se

asnicarf_data <- dplyr::as_tibble(assays(asnicarf_se)$relative_abundance, 
                         rownames = "TAXA")
asnicarf_data
```

We need to separate the "TAXA" column into seven (one per taxonomic rank) and
only keep the column containing the names of the genera:


```{r, message=FALSE, warning=FALSE}
cols <- c("KINGDOM", "PHYLUM", "CLASS", "ORDER", "FAMILY", "GENUS", "SPECIES")

asnicarf_genus <- asnicarf_data %>% 
  separate(col = TAXA, into = cols, sep = "\\|\\w__") %>% 
  mutate(KINGDOM = gsub("^k__", "", KINGDOM)) %>%
  filter(!is.na(GENUS), is.na(SPECIES)) %>% 
  select(-KINGDOM, -PHYLUM, -CLASS, -ORDER, -FAMILY, -SPECIES)

asnicarf_genus
```

To add bugphyzz attributes, we need to match the genera names in asnicarf_genus
to those of the aerophilicity data set. We can do so through the NCBI taxids.

We can add some NCBI taxids to the asnicarf_genus data with `taxize::get_uid`.

```{r, message=FALSE, warning=FALSE}
NCBI_ID <- sapply(asnicarf_genus$GENUS,  
                  function(x) taxize::get_uid(x, messages =FALSE, ask = FALSE), 
                  USE.NAMES = FALSE)

asnicarf_genus$NCBI_ID <- NCBI_ID

asnicarf_genus <- asnicarf_genus %>% 
  relocate("NCBI_ID") %>% 
  select(-GENUS) %>% 
  filter(!is.na(NCBI_ID))

asnicarf_genus
```

Now, we can combine the two data sets with the `get_attributes_abundance`
function:

```{r}
asnicarf_attributes <- get_attributes_abundance(asnicarf_genus, aerophilicity)
asnicarf_attributes

```

Exploring the data with a boxplot:

```{r}
asnicarf_attributes_long <- asnicarf_attributes %>%
  pivot_longer(cols = 2:last_col(),
               names_to = "Samples", values_to = "Abundance") 

asnicarf_attributes_long %>% 
  ggplot(aes(Attribute, Abundance)) +
  geom_boxplot()
```

Add some metadata to make the plot a bit more informative:

```{r}
asnicarf_metadata <- as_tibble(colData(asnicarf_se), rownames = "Samples")

asnicarf_attributes_long %>% 
  left_join(asnicarf_metadata, by = "Samples") %>% 
  drop_na() %>% 
  ggplot(aes(body_site, Abundance)) +
  geom_boxplot() +
  facet_wrap(~Attribute)

```

# V35 from HMP16SData

Second exemplo, with data from HMP16SData:

```{r, message=FALSE, warning=FALSE}
v35se <- V35()
v35se
```

Get features metadata and count data, and merge them into a single data set:

```{r}
v35_row_data <- rowData(v35se) %>% 
    as_tibble(rownames = "OTU")

v35_count_data <- assays(v35se)[["16SrRNA"]] %>% 
    as_tibble(rownames = "OTU")

v35_data <- full_join(v35_row_data, v35_count_data, by = "OTU")
dim(v35_data)
```

Get count data frame

```{r}
count_data <- v35_data[,8:length(v35_data)] %>% # Column 8 is GENUS; columns 9-end are samples
    filter(!is.na(GENUS)) %>% 
    group_by(GENUS) %>% 
    summarise_all(sum) %>% 
    ungroup()
dim(count_data)
count_data[1:10, 1:10]
```

Add NCBI taxids:

```{r}
NCBI_ID_v35 <- sapply(count_data$GENUS,  
                  function(x) taxize::get_uid(x, messages =FALSE, ask = FALSE), 
                  USE.NAMES = FALSE)

count_data$NCBI_ID <- NCBI_ID_v35

count_data_taxid <- count_data %>%
  drop_na() %>% 
  select(-GENUS) %>% 
  relocate("NCBI_ID") %>% 
  group_by(NCBI_ID) %>% # in case there are duplicates in HMP16SData
  summarise_all(sum) %>%
  ungroup()

# duplicated(count_data_taxid$NCBI_ID) %>% 
    # sum()
dim(count_data_taxid)
count_data_taxid[1:10,1:10]

```

Convert to a matrix, so `scale` can be used to calculate the relative abundance:

```{r}
count_matrix <- as.matrix(count_data_taxid[,-1])
rownames(count_matrix) <- count_data_taxid$NCBI_ID

# Some samples (columns) were totally empty:
# colnames(count_matrix)[colSums(count_matrix) == 0]
# "700098445" "700109123" "700111521"

# Remove empty columns
count_matrix <- count_matrix[, colnames(count_matrix)[colSums(count_matrix) > 0]]

# Calculate relative abundance
abundance <- scale(count_matrix, center = FALSE, 
               scale = colSums(count_matrix))

# Conver to a tibble in order to be processed by bugphyzz's get_score function
relative_abundance <- as_tibble(abundance, rownames = "NCBI_ID")

```

Get score matrix

```{r, message=FALSE}
score_table <- get_attributes_abundance(relative_abundance, aerophilicity)
dim(score_table)
score_table[1:10, 1:10]

# readr::write_tsv(score_matrix, file = "/home/usuario/sandbox/R/Bioconductor/score_matrix.tsv")
```

Convert table to tidy format and combine with metadata from the V35:

```{r}

df <- pivot_longer(score_table, names_to = "Samples", 
             values_to = "Abundance", cols = c(2:ncol(score_table)))

sample_metadata <- tibble::as_tibble(colData(v35se), rownames = "Samples")

final_dataset <- right_join(sample_metadata, df, by = "Samples")

```



```{r, fig.height=12}

final_dataset %>% 
    filter(Attribute %in% c("aerobic", "anaerobic", "facultatively anaerobic")) %>%
    ggplot(aes(HMP_BODY_SUBSITE, Abundance)) +
    geom_boxplot(aes(fill = HMP_BODY_SITE)) +
    facet_wrap(~Attribute, ncol = 1) +
    theme(
        axis.text.x = element_text(angle = 45, hjust = 1)
    )

```
